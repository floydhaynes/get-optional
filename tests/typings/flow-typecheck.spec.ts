import {execFile} from 'child_process';
import * as path from 'path';
import * as fs from 'fs';

const executableExtension = process.platform === 'win32' ? '.cmd' : '';
const testbed = path.join(__dirname, 'testbed');

const checkFlow = (testdir: string): Promise<string> => new Promise((resolve, reject) => {
	try {
		execFile(
			path.join(ROOT_DIR, 'node_modules/.bin/flow' + executableExtension),
			`check ${testdir} --all`.split(' '),
			{
				cwd: path.join(testbed),
			},
			(error, stdout, stderr) => {
				if (error) {
					reject(stdout || stderr || error);
				} else {
					resolve(stdout);
				}
			},
		);
	} catch (err) {
		console.error(err);
		reject(err);
	}
});

describe('Flow type definitions', () => {
	// Flow can sometimes be a bit slow
	const flowTestTimeout = 60 * 1000;

	it('should typecheck for correct usage', () => {
		expect.assertions(1);
		const result = checkFlow('flow-should-pass').then(_ => _.trim());
		return expect(result).resolves.toBe('Found 0 errors');
	}, flowTestTimeout);

	it('should not typecheck for incorrect usage', () => {
		expect.assertions(1);
		const result = checkFlow('flow-should-not-pass');
		return expect(result).rejects.toMatchSnapshot();
	}, flowTestTimeout);
});
