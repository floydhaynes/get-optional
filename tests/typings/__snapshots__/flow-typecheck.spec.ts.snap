// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flow type definitions should not typecheck for incorrect usage 1`] = `
"Error --------------------------------------------------------------------------------- flow-should-not-pass/get.js:7:14

Cannot assign \`get(...)\` to \`c\` because undefined [1] is incompatible with \`C\` [2].

   flow-should-not-pass/get.js:7:14
    7| const c: C = get(input, 'a', 'b', 'c');
                    ^^^^^^^^^^^^^^^^^^^^^^^^^

References:
   node_modules/get-optional/index.js.flow:21:5
   21| 		: void | $ElementType<$ReadOnly<$NonMaybeType<$ElementType<$ReadOnly<$NonMaybeType<$ElementType<$ReadOnly<$NonMaybeType<T>>, Key1>>>, Key2>>>, Key3>;
           ^^^^ [1]
   flow-should-not-pass/get.js:7:10
    7| const c: C = get(input, 'a', 'b', 'c');
                ^ [2]


Error -------------------------------------------------------------------------------- flow-should-not-pass/get.js:10:11

Cannot call \`get\` because:
 - Either no more than 2 arguments are expected by function type [1].
 - Or property \`c\` is missing in \`A\` [2].

   flow-should-not-pass/get.js:10:11
   10| const b = get(input, 'a', 'c');
                 ^^^^^^^^^^^^^^^^^^^^

References:
   node_modules/get-optional/index.js.flow:4:2
        v--
    4| 	<T,
    5| 		Key1: $Keys<$ReadOnly<$NonMaybeType<T>>>,
    6| 	>
    7| 		(object: T, key1: Key1)
    8| 		: void | $ElementType<$ReadOnly<$NonMaybeType<T>>, Key1>;
       		-------------------------------------------------------^ [1]
   flow-interfaces.js:21:5
   21| 	a: A;

           ^ [2]


Error -------------------------------------------------------------------------------- flow-should-not-pass/get.js:15:28

Cannot call \`get\` because:
 - Either no more than 2 arguments are expected by function type [1].
 - Or array type [2] is not an object.

   flow-should-not-pass/get.js:15:28
   15| const num: void | number = get(pollResults, 'Helsinki', 0);
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

References:
   node_modules/get-optional/index.js.flow:4:2
        v--
    4| 	<T,
    5| 		Key1: $Keys<$ReadOnly<$NonMaybeType<T>>>,
    6| 	>
    7| 		(object: T, key1: Key1)
    8| 		: void | $ElementType<$ReadOnly<$NonMaybeType<T>>, Key1>;
       		-------------------------------------------------------^ [1]
   flow-should-not-pass/get.js:14:43
   14| declare var pollResults: {[area: string]: number[]};
                                                 ^^^^^^^^ [2]


Error -------------------------------------------------------------------------------- flow-should-not-pass/get.js:20:57

Cannot call \`get\` because:
 - Either number [1] is not an object.
 - Or number [1] is not an object.
 - Or number [1] is not an object.
 - Or number [1] is not an object.
 - Or number [1] is not an object.

   flow-should-not-pass/get.js:20:57
   20| const toFixed: void | typeof Number.prototype.toFixed = get(e, 'toFixed');
                                                               ^^^^^^^^^^^^^^^^^

References:
   flow-interfaces.js:3:10
    3| type E = number;

                ^^^^^^ [1]


Error -------------------------------------------------------------------------------- flow-should-not-pass/get.js:25:24

Cannot assign \`get(...)\` to \`fromNull\` because undefined [1] is incompatible with null [2].

   flow-should-not-pass/get.js:25:24
   25| const fromNull: null = get(null, 'a');
                              ^^^^^^^^^^^^^^

References:
   node_modules/get-optional/index.js.flow:8:5
    8| 		: void | $ElementType<$ReadOnly<$NonMaybeType<T>>, Key1>;
           ^^^^ [1]
   flow-should-not-pass/get.js:25:17
   25| const fromNull: null = get(null, 'a');
                       ^^^^ [2]


Error ---------------------------------------------------------------------- flow-should-not-pass/getWithDefault.js:7:14

Cannot assign \`getWithDefault(...)\` to \`c\` because:
 - property \`arr\` is missing in \`B\` [1] but exists in \`C\` [2].
 - property \`d\` is missing in \`B\` [1] but exists in \`C\` [2].
 - property \`readonlyArr\` is missing in \`B\` [1] but exists in \`C\` [2].

   flow-should-not-pass/getWithDefault.js:7:14
   7| const c: C = getWithDefault(bDefault, input, 'a', 'b', 'c');

                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

References:
   flow-should-not-pass/getWithDefault.js:9:23
   9| declare var bDefault: B;

                            ^ [1]
   flow-should-not-pass/getWithDefault.js:7:10
   7| const c: C = getWithDefault(bDefault, input, 'a', 'b', 'c');

               ^ [2]


Error --------------------------------------------------------------------- flow-should-not-pass/getWithDefault.js:11:11

Cannot call \`getWithDefault\` because:
 - Either no more than 3 arguments are expected by function type [1].
 - Or property \`c\` is missing in \`A\` [2].

   flow-should-not-pass/getWithDefault.js:11:11
   11| const b = getWithDefault(bDefault, input, 'a', 'c');

                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

References:
   node_modules/get-optional/index.js.flow:44:2
        v-------------
   44| 	<DefaultValue,
   45| 		T,
   46| 		Key1: $Keys<$ReadOnly<$NonMaybeType<T>>>,
   47| 	>
   48| 		(defaultValue: DefaultValue, object: T, key1: Key1)
   49| 		: DefaultValue | $ElementType<$ReadOnly<$NonMaybeType<T>>, Key1>;
       		---------------------------------------------------------------^ [1]
   flow-interfaces.js:21:5
   21| 	a: A;

           ^ [2]


Error -------------------------------------------------------------------------------- flow-should-not-pass/nth.js:11:35

Cannot assign \`nth(...)\` to \`fromNullableList\` because undefined [1] is incompatible with boolean [2].

   flow-should-not-pass/nth.js:11:35
   11| const fromNullableList: boolean = nth(get(input, 'a', 'b', 'c', 'arr'), 0);

                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

References:
   node_modules/get-optional/index.js.flow:88:70
   88| declare export function nth<T>(list: ?$ReadOnlyArray<T>, n: number): void | T;
                                                                            ^^^^ [1]
   flow-should-not-pass/nth.js:11:25
   11| const fromNullableList: boolean = nth(get(input, 'a', 'b', 'c', 'arr'), 0);

                               ^^^^^^^ [2]


Error -------------------------------------------------------------------------------- flow-should-not-pass/nth.js:14:53

Cannot call \`nth\` with \`get(...)\` bound to \`list\` because boolean [1] is incompatible with string [2] in array element.

   flow-should-not-pass/nth.js:14:53
   14| const fromNullableReadonlyList: void | string = nth(get(input, 'a', 'b', 'c', 'readonlyArr'), 0);

                                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

References:
   flow-interfaces.js:10:30
   10| 	readonlyArr: $ReadOnlyArray<boolean>;

                                    ^^^^^^^ [1]
   flow-should-not-pass/nth.js:14:40
   14| const fromNullableReadonlyList: void | string = nth(get(input, 'a', 'b', 'c', 'readonlyArr'), 0);

                                              ^^^^^^ [2]


Error -------------------------------------------------------------------------------- flow-should-not-pass/nth.js:17:24

Cannot call \`nth\` with \`get(...)\` bound to \`list\` because \`B\` [1] is incompatible with read-only array type [2].

   flow-should-not-pass/nth.js:17:24
   17| const fromObject = nth(get(input, 'a', 'b'), 0);

                              ^^^^^^^^^^^^^^^^^^^^

References:
   flow-interfaces.js:17:5
   17| 	b: B;

           ^ [1]
   node_modules/get-optional/index.js.flow:88:39
   88| declare export function nth<T>(list: ?$ReadOnlyArray<T>, n: number): void | T;
                                             ^^^^^^^^^^^^^^^^^ [2]


Error -------------------------------------------------------------------------------- flow-should-not-pass/nth.js:20:40

Cannot call \`nth\` with \`'0'\` bound to \`n\` because string [1] is incompatible with number [2].

   flow-should-not-pass/nth.js:20:40
   20| const withStringIndex = nth([1, 2, 3], '0');

                                              ^^^ [1]

References:
   node_modules/get-optional/index.js.flow:88:61
   88| declare export function nth<T>(list: ?$ReadOnlyArray<T>, n: number): void | T;
                                                                   ^^^^^^ [2]



Found 13 errors

Only showing the most relevant union/intersection branches.
To see all branches, re-run Flow with --show-all-branches
"
`;
